src/IO/H5/File.hpp

```diff
@@ -0,0 +1,328 @@
+  template <typename ObjectType>
+  const ObjectType& convert_to_derived(
+      const std::unique_ptr<h5::Object>& current_object) const;
+
+  template <AccessType Access_lt = Access_t,
+            std::enable_if_t<Access_lt == AccessType::ReadWrite>* = nullptr>
+  void insert_header() {
+    insert<h5::Header>("/header");
+  }
+  template <AccessType Access_lt = Access_t,
+            std::enable_if_t<Access_lt == AccessType::ReadOnly>* = nullptr>
+  void insert_header() {}  // LCOV_EXCL_LINE
+
+  template <typename ObjectType>
+  std::tuple<bool, detail::OpenGroup, std::string> check_if_object_exists(
+      const std::string& in_name) const;
+
+  std::string file_name_;
+  hid_t file_id_{-1};
+  mutable std::unique_ptr<h5::Object> current_object_{nullptr};
+  std::vector<std::string> h5_groups_;
+  /// \endcond HIDDEN_SYMBOLS
+};
+
+// ======================================================================
+// H5File Definitions
+// ======================================================================
+
+template <AccessType Access_t>
+template <typename ObjectType, typename... Args,
+          typename std::enable_if_t<(sizeof(ObjectType),
+                                     Access_t == AccessType::ReadWrite)>*>
+ObjectType& H5File<Access_t>::get(const std::string& path_and_name,
+                                  Args&&... args) {
+  try {
+    current_object_ = nullptr;
+    // C++17: structured bindings
+    auto exists_group_name = check_if_object_exists<ObjectType>(path_and_name);
+    hid_t group_id = std::get<1>(exists_group_name).id();
+    current_object_ = std::make_unique<ObjectType>(
+        std::get<0>(exists_group_name),
+        std::move(std::get<1>(exists_group_name)), group_id,
+        std::move(std::get<2>(exists_group_name)), std::forward<Args>(args)...);
+    if (not std::get<0>(exists_group_name)) {
+      // We need to close the H5 file otherwise we get the error:
+      //
+      // pure virtual method called
+      // terminate called recursively
+      current_object_ = nullptr;
+      CHECK_H5(H5Fclose(file_id_),
+               "Failed to close file: '" << file_name_ << "'");
+      ERROR("Cannot open the object '"
+            << path_and_name + ObjectType::extension()
+            << "' because it does not exist.");
+    }
+    return dynamic_cast<ObjectType&>(*current_object_);
+    // LCOV_EXCL_START
+  } catch (const std::bad_cast& e) {
+    current_object_ = nullptr;
+    CHECK_H5(H5Fclose(file_id_),
+             "Failed to close file: '" << file_name_ << "'");
+    ERROR("Failed to cast " << path_and_name << " to "
+                            << pretty_type::get_name<ObjectType>()
+                            << ".\nCast error: " << e.what());
+  } catch (const std::exception& e) {
+    current_object_ = nullptr;
+    CHECK_H5(H5Fclose(file_id_),
+             "Failed to close file: '" << file_name_ << "'");
+    ERROR("Unknown exception caught: " << e.what());
+    // LCOV_EXCL_STOP
+  }
+}
```

Comment:
if the const and non-const versions are the same, please don't duplicate the code